task_manager/task_manager.hh

```diff
@@ -0,0 +1,229 @@
+                    return child->get_progress();
+                });
+            }, progress{}, [] (progress l, progress r) {
+                progress res{};
+                res.total = l.total + r.total;
+                res.completed = l.completed + r.completed;
+                return res;
+            });
+
+        }
+
+        virtual std::string get_module() const = 0;
+
+        virtual is_abortable is_abortable() const {
+            return is_abortable::no;
+        }
+
+        void set_number(int number) {
+            _status.number = number;
+        }
+
+        void set_type(std::string type) {
+            _status.type = type;
+        }
+
+        void change_state(task_state state) {
+            _status.state = state;
+        }
+
+        void set_keyspace(std::string keyspace) {
+            _status.keyspace = keyspace;
+        }
+
+        void set_table(std::string table) {
+            _status.table = table;
+        }
+
+        void set_entity(std::string entity) {
+            _status.entity = entity;
+        }
+
+        void set_total_progress(double progress) {
+            _status.progress.total = progress;
+        }
+
+        void report_completion_progress(double progress) {
+            _status.progress.completed += progress;
+        }
+
+        void start(task_manager& tm) {
+            _status.start_time = db_clock::now();
+            _status.state = task_manager::task_state::running;
+
+            (void)with_gate(tm.async_gate(), [f = done(), &tm, id = _status.id] () mutable {
+                return std::move(f).finally([&tm, id] {
+                    return sleep_abortable(tm.task_ttl(), tm.abort_source());
+                }).finally([&tm, id] {
+                    tm.unregister_task(id);
+                });
+            });
+        }
+
+        virtual task_ptr create_child(task_id uuid, foreign_task_ptr&& parent) const = 0;
+
+        future<> start_on_shards(task_manager& tm, std::vector<task_id>& uuids, task_ptr this_task) {
+            return tm.container().map([parent = seastar::make_foreign(this_task), &uuids] (task_manager& tm) mutable {
+                task_ptr task = parent->create_child(uuids[this_shard_id()], std::move(parent));
+                tm.register_task(task);
+                return seastar::make_foreign<task_ptr>(task);
+            }).then([this, &tm] (foreign_task_vector tasks) {
+                this->_children = std::move(tasks);
+                start(tm);
+            });
+        }
+
+        future<> done() {
+            return _done.get_shared_future();
+        }
+
+        void finish() {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::done;
+            _done.set_value();
+        }
+
+        void finish_failed(std::exception&& ex) {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::failed;
+            _status.error = ex.what();
```

Comment:
Better get a std::exception_ptr plus an optional error message as input.
If the error parameter is not given, we can generate it here using `fmt::format("{}", ex)`.

In any case, we usually prefer passing std::exception_ptr around rather than `std::exception` references.